homeassistant/components/hassmpris/__init__.py

```diff
@@ -0,0 +1,102 @@
+"""The MPRIS media playback remote control integration."""
+from __future__ import annotations
+
+from typing import Any, cast
+
+from cryptography.hazmat.primitives.asymmetric.rsa import RSAPrivateKey
+import hassmpris_client
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import EVENT_HOMEASSISTANT_STOP, Platform
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryAuthFailed, ConfigEntryNotReady
+
+from .cert_data import CertStore
+from .const import CONF_HOST, CONF_MPRIS_PORT, DOMAIN, LOGGER as _LOGGER
+from .models import ConfigEntryData, HassmprisData
+
+PLATFORMS: list[Platform] = [Platform.MEDIA_PLAYER]
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up MPRIS media playback remote control from a config entry."""
+
+    # This would be an error, and the type checker knows it.
+    assert entry.unique_id is not None
+
+    try:
+        client_cert, client_key, trust_chain = await CertStore(
+            hass, entry.unique_id
+        ).load_cert_data()
+    except KeyError as exc:
+        raise ConfigEntryAuthFailed(
+            "Authentication data is missing -- must reauth"
+        ) from exc
+
+    entry_data = cast(ConfigEntryData, entry.data)
+    clnt = hassmpris_client.AsyncMPRISClient(
+        entry_data[CONF_HOST],
+        entry_data[CONF_MPRIS_PORT],
+        client_cert,
+        cast(RSAPrivateKey, client_key),
+        trust_chain,
+    )
+    try:
+        _LOGGER.debug("Pinging the MPRIS agent")
+        await clnt.ping()
+        _LOGGER.info("Successfully pinged the MPRIS agent")
+
+    except hassmpris_client.Unauthenticated as exc:
+        raise ConfigEntryAuthFailed(exc) from exc
+    except Exception as exc:
+        raise ConfigEntryNotReady(str(exc)) from exc
+
+    hass.data.setdefault(DOMAIN, {})
+    component_data = hass.data[DOMAIN][entry.entry_id] = HassmprisData(
+        clnt, [], None, None
+    )
+
+    async def _run_unloaders(*unused_args: Any) -> None:
+        _LOGGER.debug("Running unloaders")
+        while component_data.unloaders:
+            await component_data.unloaders.pop()()
+        _LOGGER.debug("Unloaders complete")
+
+    component_data.unload_func = _run_unloaders
+
+    entry.async_on_unload(
+        hass.bus.async_listen_once(
+            EVENT_HOMEASSISTANT_STOP,
+            component_data.unload_func,
+        )
+    )
```

Comment:
This all seems pointless / over complicated.
Instead of adding functions to `unloaders`, why not just add them to `entry.async_on_unload`?
Or do we need to ensure that the unloaders run when Home Assistant is shutting down, and if that's the case, why?

Reply:
This integration cleans everything up on HA shutdown, disconnecting from the server and updating the state of the entities to unavailable.

If I skip the execution of the cleanup routines, I get a bunch of exceptions at HA exit time, since this integration is still running and nothing has stopped the loop.  Let me show you what HA shows on exit when we do not actually clean up the loop:

```
2023-12-19 05:01:13.377 ERROR (MainThread) [homeassistant.components.hassmpris] We lost connectivity after 3 updates (Server gone: Final process shutdown (<class 'asyncio.exceptions.CancelledError'>)) -- reconnecting
Traceback (most recent call last):
  File "/home/user/Projects/homeassistant-core/venv/lib64/python3.11/site-packages/hassmpris_client/__init__.py", line 129, in inner
    async for x in f(*args, **kwargs):
  File "/home/user/Projects/homeassistant-core/venv/lib64/python3.11/site-packages/hassmpris_client/__init__.py", line 373, in stream_updates
    message = await stream.recv_message()
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/Projects/homeassistant-core/venv/lib64/python3.11/site-packages/grpclib/client.py", line 428, in recv_message
    message = await recv_message(self._stream, self._codec,
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/Projects/homeassistant-core/venv/lib64/python3.11/site-packages/grpclib/stream.py", line 20, in recv_message
    meta = await stream.recv_data(5)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/Projects/homeassistant-core/venv/lib64/python3.11/site-packages/grpclib/protocol.py", line 347, in recv_data
    return await self.buffer.read(size)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/Projects/homeassistant-core/venv/lib64/python3.11/site-packages/grpclib/protocol.py", line 88, in read
    data, data_size, ack_size = await self._unacked.get()
                                ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/asyncio/queues.py", line 158, in get
    await getter
asyncio.exceptions.CancelledError: Final process shutdown

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/user/Projects/homeassistant-core/homeassistant/components/hassmpris/media_player.py", line 425, in run
    async for _ in self._monitor_updates():
  File "/home/user/Projects/homeassistant-core/homeassistant/components/hassmpris/media_player.py", line 499, in _monitor_updates
    async for update in self.client.stream_updates(
  File "/home/user/Projects/homeassistant-core/venv/lib64/python3.11/site-packages/hassmpris_client/__init__.py", line 140, in inner
    raise Disconnected(e) from e
hassmpris_client.Disconnected: Server gone: Final process shutdown (<class 'asyncio.exceptions.CancelledError'>)
2023-12-19 05:01:13.405 DEBUG (MainThread) [homeassistant.components.hassmpris] Manually updated MPRIS data
2023-12-19 05:01:13.406 DEBUG (MainThread) [homeassistant.components.hassmpris] VLC media player: Now unavailable
2023-12-19 05:01:13.406 DEBUG (MainThread) [homeassistant.components.hassmpris] VLC media player: Writing HA state
2023-12-19 05:01:13.406 DEBUG (MainThread) [homeassistant.components.hassmpris] Videos: Now unavailable
2023-12-19 05:01:13.406 DEBUG (MainThread) [homeassistant.components.hassmpris] Videos: Writing HA state
2023-12-19 05:01:13.406 DEBUG (MainThread) [homeassistant.components.hassmpris] :1.753: Now unavailable
2023-12-19 05:01:13.406 DEBUG (MainThread) [homeassistant.components.hassmpris] :1.753: Writing HA state
2023-12-19 05:01:13.406 DEBUG (MainThread) [homeassistant.components.hassmpris] :1.768: Now unavailable
2023-12-19 05:01:13.406 DEBUG (MainThread) [homeassistant.components.hassmpris] :1.768: Writing HA state
2023-12-19 05:01:18.385 WARNING (MainThread) [homeassistant.runner] Task could not be canceled and was still running after shutdown: <Task cancelling name='Task-2048' coro=<MPRISCoordinator.run() running at /home/user/Projects/homeassistant-core/homeassistant/components/hassmpris/media_player.py:425> wait_for=<Future pending cb=[Task.task_wakeup()]>>
2023-12-19 05:01:18.387 DEBUG (MainThread) [homeassistant.components.hassmpris] Setting absent player VLC media player to off
2023-12-19 05:01:18.387 DEBUG (MainThread) [homeassistant.components.hassmpris] Setting absent player Videos to off
2023-12-19 05:01:18.387 DEBUG (MainThread) [homeassistant.components.hassmpris] Setting absent player :1.753 to off
2023-12-19 05:01:18.387 DEBUG (MainThread) [homeassistant.components.hassmpris] Setting absent player :1.768 to off
2023-12-19 05:01:18.387 ERROR (MainThread) [homeassistant] Error doing job: an error occurred during closing of asynchronous generator <async_generator object MPRISCoordinator._monitor_updates at 0x7f847c5e7740>
RuntimeError: aclose(): asynchronous generator is already running
2023-12-19 05:01:18.387 ERROR (MainThread) [homeassistant] Error doing job: an error occurred during closing of asynchronous generator <async_generator object AsyncMPRISClient.stream_updates at 0x7f8485730840>
RuntimeError: aclose(): asynchronous generator is already running
2023-12-19 05:01:18.387 ERROR (MainThread) [homeassistant] Error doing job: an error occurred during closing of asynchronous generator <async_generator object AsyncMPRISClient.stream_updates at 0x7f847c26d440>
RuntimeError: aclose(): asynchronous generator is already running
```

Is this not how we do things?

Leaving this comment open and unresolved in the hopes of guidance here.

Reply:
To clarify:

`on_unload()` runs the registered unloaders before disconnecting the client.

The first unloader registered is the `EntityCoordinator.stop()` method.

The `stop()` method is called first.  Then the client is disconnected.

If we do not disconnect the client, HA will complain about coroutines still running as it exits.  So we must disconnect the client.

If we do not call `stop()` on the coordinator _before_ disconnecting the client, HA will complain loudly with an exception as it exits because all the coordinator sees is a connection error (after disconnecting the client).

This is why the unload process is structured the way it is.

Reply:
For the record: I *just tried* using the `config_entry.async_on_unload` system to make things how you suggested them, such that the coordinator is shut down that way.  It did not work — it only worked in the HA shutdown case, but not in the reload case.  I got the exceptions because the unload tasks happen *in parallel*, and this shutdown process *requires strict ordering* (shut down coordinator, then disconnect from client),

Reply:
Please let me know what you'd like to see in order to consider this definitively resolved.  Thanks!